packages/babel-generator/src/buffer.js

```diff
@@ -1,275 +1,121 @@
-
-    if (this.buf && !this.endsWith(" ") && !this.endsWith("\n")) {
-      this.push(" ");
-    }
+  removeTrailingNewline(): void {
+    if (this._queue.length > 0 && this._queue[0][0] === "\n") this._queue.shift();
   }
 
-  /**
-   * Writes a token that can't be safely parsed without taking whitespace into account.
-   */
-
-  word(str: string) {
-    if (this._endsWithWord) this.push(" ");
-
-    this.push(str);
-    this._endsWithWord = true;
+  removeLastSemicolon(): void {
+    if (this._queue.length > 0 && this._queue[0][0] === ";") this._queue.shift();
   }
 
-  /**
-   * Writes a simple token.
-   */
-
-  token(str: string) {
-    // space is mandatory to avoid outputting <!--
-    // http://javascript.spec.whatwg.org/#comment-syntax
-    if ((str === "--" && this.last === "!") ||
-
-      // Need spaces for operators of the same kind to avoid: `a+++b`
-      (str[0] === "+" && this.last === "+") ||
-      (str[0] === "-" && this.last === "-")) {
-      this.push(" ");
+  endsWith(str: string): boolean {
+    const end = this._last + this._queue.reduce((acc, item) => item[0] + acc, "");
+    if (str.length <= end.length) {
+      return end.slice(-str.length) === str;
     }
 
-    this.push(str);
-  }
-
-
-  /**
-   * Remove the last character.
-   */
-
-  removeLast(cha: string) {
-    if (this.format.compact) return;
-    return this._removeLast(cha);
+    // We assume that everything being matched is at most a single token plus some whitespace,
+    // which everything currently is, but otherwise we'd have to expand _last or check _buf.
+    return false;
   }
 
-  _removeLast(cha: string) {
-    if (!this.endsWith(cha)) return;
-    this.buf = this.buf.slice(0, -1);
-    this.last = this.buf[this.buf.length - 1];
-    this.position.unshift(cha);
-  }
-
-  /**
-   * Set some state that will be modified if a newline has been inserted before any
-   * non-space characters.
-   *
-   * This is to prevent breaking semantics for terminatorless separator nodes. eg:
-   *
-   *    return foo;
-   *
-   * returns `foo`. But if we do:
-   *
-   *   return
-   *   foo;
-   *
-   *  `undefined` will be returned and not `foo` due to the terminator.
-   */
-
-  startTerminatorless(): Object {
-    return this.parenPushNewlineState = {
-      printed: false
-    };
-  }
-
-  /**
-   * Print an ending parentheses if a starting one has been printed.
-   */
-
-  endTerminatorless(state: Object) {
-    if (state.printed) {
-      this.dedent();
-      this.newline();
-      this.token(")");
-    }
-  }
-
-  /**
-   * Add a newline (or many newlines), maintaining formatting.
-   * Strips multiple newlines if removeLast is true.
-   */
-
-  newline(i?: boolean | number, removeLast?: boolean) {
-    if (this.format.retainLines || this.format.compact) return;
-
-    if (this.format.concise) {
-      this.space();
-      return;
+  getLast(): string {
+    if (this._queue.length > 0) {
+      const last = this._queue[this._queue.length - 1][0];
```

Comment:
Should this be `this._queue[0][0]`? Right now, we're checking the last char of the item queued first (since we `#unshift`), not the item most recently queued.